slotsmanager.c

```diff
@@ -127,46 +130,90 @@ slots_gc(void)
      * link lists first. likely worth it, since GC happens only in domain decompose
      * and snapshot IO, both take far more time than rebuilding the tree. */
     tree_invalid |= slots_gc_base();
-    tree_invalid |= slots_gc_slots();
+    tree_invalid |= slots_gc_slots(compact_slots);
 
     MPI_Allreduce(MPI_IN_PLACE, &tree_invalid, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
 
-    EISlotsAfterGC event = {0};
+    EISlotsAfterGC event = {{tree_invalid}};
 
     event_emit(&EventSlotsAfterGC, (EIBase*) &event);
 
     return tree_invalid;
 }
 
+
+#define GARBAGE(i, ptype) (ptype >= 0 ? BASESLOT_PI(i,ptype)->IsGarbage : P[i].IsGarbage)
+#define PART(i, ptype) (ptype >= 0 ? (void *) BASESLOT_PI(i, ptype) : (void *) &P[i])
+
+/*Compaction algorithm*/
+static int
+slots_gc_compact(int used, int ptype, size_t size)
```

Comment:
It feels a bit strange using ptype < 0 for the base slots -- I sort of felt baseslots are different from secondary slots. Perhaps we shall have both inheriting from GCFriendly, which is a 4 bytes or 8 byte struct with IsGarbage flag?

Reply:
This is better to be left on a future PR though.

Reply:
Yes - let's leave it like this for now and make it a future PR if we find ourselves doing this in more than one place.